lib/libvfio-user.c

```diff
@@ -927,126 +901,147 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
     }
     return 0;
 }
 
+static uint32_t
+device_feature_flags(uint32_t feature) {
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_MIGRATION:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT:
+            return VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE:
+            return VFIO_DEVICE_FEATURE_GET
+                | VFIO_DEVICE_FEATURE_SET
+                | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_START:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP:
+            return VFIO_DEVICE_FEATURE_SET | VFIO_DEVICE_FEATURE_PROBE;
+        default:
+            return 0;
+    };
+}
+
 static int
-handle_dirty_pages_get(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
 {
-    struct vfio_user_dirty_pages *dirty_pages_in;
-    struct vfio_user_dirty_pages *dirty_pages_out;
-    struct vfio_user_bitmap_range *range_in;
-    struct vfio_user_bitmap_range *range_out;
-    size_t argsz;
-    int ret;
+    assert(vfu_ctx != NULL);
+    assert(msg != NULL);
 
+    if (vfu_ctx->migration == NULL ||
+        msg->in.iov.iov_len < sizeof(struct vfio_user_device_feature)) {
+        return ERROR_INT(EINVAL);
+    }
 
-    dirty_pages_in = msg->in.iov.iov_base;
+    struct vfio_user_device_feature *req = msg->in.iov.iov_base;
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages_in) + sizeof(*range_in) ||
-        dirty_pages_in->argsz > SERVER_MAX_DATA_XFER_SIZE ||
-        dirty_pages_in->argsz < sizeof(*dirty_pages_out)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size=%zu argsz=%u",
-                msg->in.iov.iov_len, dirty_pages_in->argsz);
+    uint32_t supported_flags =
+        device_feature_flags(req->flags & VFIO_DEVICE_FEATURE_MASK);
+
+    if ((req->flags & supported_flags) !=
+        (req->flags & ~VFIO_DEVICE_FEATURE_MASK) || supported_flags == 0) {
         return ERROR_INT(EINVAL);
     }
 
-    range_in = msg->in.iov.iov_base + sizeof(*dirty_pages_in);
+    uint32_t feature = req->flags & VFIO_DEVICE_FEATURE_MASK;
 
-    /*
-     * range_in is client-controlled, but we only need to protect against
-     * overflow here: we'll take MIN() against a validated value next, and
-     * dma_controller_dirty_page_get() will validate the actual ->bitmap.size
-     * value later, anyway.
-     */
-    argsz = satadd_u64(sizeof(*dirty_pages_out) + sizeof(*range_out),
-                       range_in->bitmap.size);
+    ssize_t ret;
 
-    msg->out.iov.iov_len = MIN(dirty_pages_in->argsz, argsz);
-    msg->out.iov.iov_base = malloc(msg->out.iov.iov_len);
-    if (msg->out.iov.iov_base == NULL) {
-        return -1;
-    }
-    dirty_pages_out = msg->out.iov.iov_base;
-    memcpy(dirty_pages_out, dirty_pages_in, sizeof(*dirty_pages_out));
-    dirty_pages_out->argsz = argsz;
+    if (req->flags & VFIO_DEVICE_FEATURE_PROBE) {
+        msg->out.iov.iov_base = malloc(msg->in.iov.iov_len);
```

Comment:
I don't think you've verified this iov_len is a safe value

Reply:
`msg->in.iov.iov_len` is safe, it's checked when the message is loaded in (in `in_valid_header`).